components/common/src/command/package/download.rs

```diff
@@ -0,0 +1,384 @@
+                         Error::APIError,
+                         Package},
+            hcore::{self,
+                    crypto::{artifact,
+                             keys::parse_name_with_rev,
+                             SigKeyPair},
+                    fs::{cache_artifact_path,
+                         cache_key_path},
+                    package::{PackageArchive,
+                              PackageIdent,
+                              PackageIdentTarget,
+                              PackageTarget},
+                    ChannelIdent}};
+
+use reqwest::StatusCode;
+use retry::{delay,
+            retry};
+
+use crate::{error::{Error,
+                    Result},
+            ui::{Status,
+                 UIWriter}};
+
+pub const RETRIES: usize = 5;
+pub const RETRY_WAIT: Duration = Duration::from_millis(3000);
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum InstallMode {
+    Online,
+    Offline,
+}
+
+impl Default for InstallMode {
+    fn default() -> Self { InstallMode::Online }
+}
+
+/// Represents a fully-qualified Package Identifier, meaning that the normally optional version and
+/// release package coordinates are guaranteed to be set. This fully-qualified-ness is checked on
+/// construction and as the underlying representation is immutable, this state does not change.
+
+/// Download a Habitat package.
+///
+/// If an `PackageIdentTarget` is given, we retrieve the package
+/// from the specified Builder `url`. Providing a fully-qualified
+/// identifer will result in that exact package being installed
+/// (regardless of `channel`). Providing a partially-qualified
+/// identifier will result in the installation of latest appropriate
+/// release from the given `channel`.
+///
+/// Any dependencies of will be retrieved from Builder (if they're not
+/// already cached locally).
+///
+/// At the end of this function, the specified package and all its
+/// dependencies will be downloaded on the system.
+
+/// Note: it's worth investigating whether
+/// LocalPackageUsage makes sense here
+/// Also, in the future we may want to accept an alternate builder to 'filter' what we pull down by
+/// That would greatly optimize the 'sync' to on prem builder case, as we could point to that
+/// and only fetch what we don't already have.
+#[allow(clippy::too_many_arguments)]
+pub fn start<U>(ui: &mut U,
+                url: &str,
+                channel: &ChannelIdent,
+                product: &str,
+                version: &str,
+                idents: Vec<PackageIdent>,
+                target: PackageTarget,
+                fs_root_path: Option<&PathBuf>,
+                token: Option<&str>)
+                -> Result<()>
+    where U: UIWriter
+{
+    debug!("Starting download with url: {}, channel: {}, product: {}, version: {}, target: {}, \
+            fs_root_path: {:?}, token: {:?}",
+           url, channel, product, version, target, fs_root_path, token);
+
+    let key_cache_path = &cache_key_path(fs_root_path);
```

Comment:
This is probably the bug we were discussing... if the caller specifies a path it should get passed in and used instead